File: Programs/Two_Sum.cpp
#include<bits/stdc++.h>
using namespace std;

/* CodingNinjas:
You are given an array of integers 'ARR' of length 'N' and an integer Target. Your task is to return all pairs of elements such that they add up to Target.
Note: We cannot use the element at a given index twice.
Follow Up: Try to do this problem in O(N) time complexity. 
 */
vector<pair<int,int>> twoSum(vector<int>& arr, int target, int n) {
    unordered_map<int, int> freq;
    vector<pair<int,int>> ans;
    for(int& num: arr)
        freq[num]++;

    for(int& num: arr) {
        int complement= target - num;

        if(freq[complement] > 0) {
            if(num == complement) {
                if(freq[num] >= 2) {
                    ans.push_back({num, complement});
                    freq[num] -= 2;
                }
            } else if(freq[num] >= 1 && freq[complement] >= 1) {
                    ans.push_back({num, complement});
                    freq[num]--;
                    freq[complement]--;
            }
        }
    }

    if(ans.empty())
        ans.push_back({-1, -1});
    return ans;
}

/* LeetCode
Given an array of integers nums and an integer target, return indices of the two numbers such that they add up to target.
You may assume that each input would have exactly one solution, and you may not use the same element twice.
You can return the answer in any order.
 */
class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        unordered_map<int, int> numToIndex;
        
        for (int i = 0; i < nums.size(); ++i) {
            int complement = target - nums[i];
            
            if (numToIndex.find(complement) != numToIndex.end()) {
                return {numToIndex[complement], i};
            }

            numToIndex[nums[i]] = i;
        }

        return {};
    }
};

File: Programs/Two_Sum.py
import List
from collections import Counter

# CodingNinjas
def twoSum(arr, target, n):
    ans= list()

    freq= dict()
    for num in arr:
        if num in freq:
            freq[num] += 1
        else:
            freq[num]= 1
    # freq= Counter(arr)

    for num in arr:
        complement= target - num

        if freq.get(complement, 0) > 0:
            if num == complement:
                if freq[num] > 1:
                    ans.append([num, num])
                    freq[num] -= 2
            elif freq[num] > 0 and freq[complement] > 0:
                ans.append([num, complement])
                freq[num] -= 1
                freq[complement] -= 1

    if not ans:
        ans.append([-1, -1])

    return ans

# LeetCode
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        numToIndex= dict()
        for index, num in enumerate(nums):
            complement= target - num

            if complement in numToIndex.keys():
                return [index, numToIndex[complement]]

            numToIndex[num]= index
        return []
